src/extensions/factory/factoryext_hooks.cpp

```diff
@@ -45,59 +119,84 @@
-			goto production_completed;
-		}
-
-		/**
-		 *  If InstantBuild is toggled on, make sure the local player is a human house.
-		 */
-		if (Vinifera_Developer_InstantBuild
-			&& this_ptr->House->Is_Human_Control() && this_ptr->House == PlayerPtr) {
-
-			this_ptr->StageClass::Set_Stage(FactoryClass::STEP_COUNT);
-			goto production_completed;
-		}
-
-		/**
-		 *  If the AI has taken control of the player house, it needs a special
-		 *  case to handle the "player" instant build mode.
-		 */
-		if (Vinifera_Developer_InstantBuild) {
-			if (Vinifera_Developer_AIControl && this_ptr->House == PlayerPtr) {
-
-				this_ptr->StageClass::Set_Stage(FactoryClass::STEP_COUNT);
-				goto production_completed;
-			}
-		}
-
-	}
-
-	/**
-	 *  Stolen bytes/code.
-	 */
-	if (this_ptr->StageClass::Fetch_Stage() == FactoryClass::STEP_COUNT) {
-		goto production_completed;
-	}
+    GET_REGISTER_STATIC(FactoryClass *, this_ptr, esi);
+
+    if (Vinifera_DeveloperMode) {
+
+        /**
+         *  If AIInstantBuild is toggled on, make sure this is a non-human AI house.
+         */
+        if (Vinifera_Developer_AIInstantBuild
+            && !this_ptr->House->Is_Human_Control() && this_ptr->House != PlayerPtr) {
+
+            this_ptr->StageClass::Set_Stage(FactoryClass::STEP_COUNT);
+            goto production_completed;
+        }
+
+        /**
+         *  If InstantBuild is toggled on, make sure the local player is a human house.
+         */
+        if (Vinifera_Developer_InstantBuild
+            && this_ptr->House->Is_Human_Control() && this_ptr->House == PlayerPtr) {
+
+            this_ptr->StageClass::Set_Stage(FactoryClass::STEP_COUNT);
+            goto production_completed;
+        }
+
+        /**
+         *  If the AI has taken control of the player house, it needs a special
+         *  case to handle the "player" instant build mode.
+         */
+        if (Vinifera_Developer_InstantBuild) {
+            if (Vinifera_Developer_AIControl && this_ptr->House == PlayerPtr) {
+
+                this_ptr->StageClass::Set_Stage(FactoryClass::STEP_COUNT);
+                goto production_completed;
+            }
+        }
+
+    }
+
+    /**
+     *  Stolen bytes/code.
+     */
+    if (this_ptr->StageClass::Fetch_Stage() == FactoryClass::STEP_COUNT) {
+        goto production_completed;
+    }
 
 function_return:
-	JMP(0x00496FA3);
+    JMP(0x00496FA3);
 
     /**
      *  Production Completed, then suspend further production.
      */
 production_completed:
-	JMP(0x00496F73);
+    JMP(0x00496F73);
+}
+
+
+/**
+ *  Update the queue and remove any items that are no
+ *  longer buildable by the factory owner's house
+ *
+ *  @author: ZivDero
+ */
+DECLARE_PATCH(_FactoryClass_AI_Abandon_If_Cant_Build)
+{
+    GET_REGISTER_STATIC(FactoryClassFake*, this_ptr, ecx);
+
+    _asm push esi
+
+    this_ptr->_Verify_Can_Build();
+
+    _asm
+    {
+        pop esi
+        mov al, [esi + 5Ch]
+        test al, al
+    }
+
+    JMP_REG(ebx, 0x00496EAC);
```

Comment:
1) There is no need to manually preserve `esi`. `esi` must be preserved in function calls in most x86 calling conventions, meaning the function call won't trash it (the compiler is not allowed to generate code in the function that would trash `esi`).

2) Doing test before unconditional jmp is a bad practice. Test instructions should be immediately followed by a conditional jump.

3) Like `esi`, `ebx` must also be preserved in function calls, meaning you aren't allowed to modify its value unless you push/pop it to preserve its original value.

4) Since we know that `esi+5Ch` is `FactoryClass.IsSuspended`, we might as well rewrite it in C++.

```suggestion
    GET_REGISTER_STATIC(FactoryClassFake*, this_ptr, ecx);

    this_ptr->_Verify_Can_Build();

    // Stolen bytes / code.
    if (this_ptr->IsSuspended) {
        // Factory is suspended, exit from function.
        JMP(0x00496FA3);
    }
    
    // Continue factory AI logic.
    JMP(0x00496EB2);
```